File: worker/lib/secrets.py
"""At-rest encryption for stored Southwest account passwords.

Counterpart of frontend/lib/secrets.ts — the format must stay in sync:

    enc:v1:base64(iv[12] || tag[16] || ciphertext)   (AES-256-GCM)

Key = SHA-256(AUTH_SECRET). The frontend encrypts when an account is saved;
the worker decrypts when it needs to log in to Southwest. Legacy plaintext
values (no prefix) pass through unchanged and are re-encrypted at startup.
"""

from __future__ import annotations

import base64
import hashlib
import os

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

PREFIX = "enc:v1:"


class SecretDecryptError(Exception):
    pass


def _get_key() -> bytes | None:
    secret = os.environ.get("AUTH_SECRET")
    if not secret:
        return None
    return hashlib.sha256(secret.encode()).digest()


def is_encrypted(value: str) -> bool:
    return isinstance(value, str) and value.startswith(PREFIX)


def encrypt_secret(plaintext: str) -> str:
    key = _get_key()
    if not key:
        return plaintext
    iv = os.urandom(12)
    # AESGCM.encrypt returns ciphertext||tag; the shared format is iv||tag||ciphertext
    ct_and_tag = AESGCM(key).encrypt(iv, plaintext.encode(), None)
    ciphertext, tag = ct_and_tag[:-16], ct_and_tag[-16:]
    return PREFIX + base64.b64encode(iv + tag + ciphertext).decode()


def decrypt_secret(value: str) -> str:
    if not is_encrypted(value):
        return value
    key = _get_key()
    if not key:
        raise SecretDecryptError("AUTH_SECRET is not set; cannot decrypt stored credential")
    raw = base64.b64decode(value[len(PREFIX):])
    iv, tag, ciphertext = raw[:12], raw[12:28], raw[28:]
    try:
        return AESGCM(key).decrypt(iv, ciphertext + tag, None).decode()
    except Exception as err:
        raise SecretDecryptError(
            "Could not decrypt stored Southwest password — AUTH_SECRET has likely "
            "changed since the account was saved. Re-enter the account password "
            "in the web UI."
        ) from err

File: frontend/lib/secrets.ts
import crypto from "crypto";

/**
 * At-rest encryption for stored Southwest account passwords.
 *
 * The worker must recover the plaintext to log in to Southwest, so a one-way
 * hash is impossible — instead values are encrypted with AES-256-GCM using a
 * key derived from AUTH_SECRET. A copied database file (a backup, a stolen
 * volume snapshot) then no longer exposes airline credentials on its own.
 *
 * Format (must stay in sync with worker/lib/secrets.py):
 *   enc:v1:base64(iv[12] || tag[16] || ciphertext)
 *
 * Note: changing AUTH_SECRET makes existing values undecryptable — Southwest
 * account passwords must be re-entered after rotating it.
 */

const PREFIX = "enc:v1:";

function getKey(): Buffer | null {
  const secret = process.env.AUTH_SECRET;
  if (!secret) return null;
  return crypto.createHash("sha256").update(secret).digest();
}

export function encryptSecret(plaintext: string): string {
  const key = getKey();
  if (!key) return plaintext; // no secret configured (dev) — store as-is
  const iv = crypto.randomBytes(12);
  const cipher = crypto.createCipheriv("aes-256-gcm", key, iv);
  const ciphertext = Buffer.concat([cipher.update(plaintext, "utf8"), cipher.final()]);
  const tag = cipher.getAuthTag();
  return PREFIX + Buffer.concat([iv, tag, ciphertext]).toString("base64");
}

export function decryptSecret(value: string): string {
  if (!value.startsWith(PREFIX)) return value; // legacy plaintext
  const key = getKey();
  if (!key) throw new Error("AUTH_SECRET is not set; cannot decrypt stored credential");
  const raw = Buffer.from(value.slice(PREFIX.length), "base64");
  const iv = raw.subarray(0, 12);
  const tag = raw.subarray(12, 28);
  const ciphertext = raw.subarray(28);
  const decipher = crypto.createDecipheriv("aes-256-gcm", key, iv);
  decipher.setAuthTag(tag);
  return Buffer.concat([decipher.update(ciphertext), decipher.final()]).toString("utf8");
}
